File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/AbortableCommand.java
package org.extensiblecommands;

import static java.util.Objects.requireNonNull;

/**
 * Implements command that can be aborted during its execution by invoking Abort() method.
 * This implies that command execution and its abort should occur in different threads.
 */
public class AbortableCommand extends DecoratorCommand {
    /**
     * Method to abort this command
     */
    private final AbortDelegate abortDelegate;

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param abortDelegate     Delegate to abort operation of the Core command
     */
    public AbortableCommand(Command coreCommand, AbortDelegate abortDelegate) {
        this(coreCommand, abortDelegate, "Abortable");
    }

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param abortDelegate     Delegate to abort operation of the Core command
     * @param name              Command name
     */
    public AbortableCommand(Command coreCommand, AbortDelegate abortDelegate, String name) {
        super(coreCommand, name);
        requireNonNull(abortDelegate, String.format("Abort Delegate is NULL in AbortableCommand %s", this.name));

        this.abortDelegate = abortDelegate;
    }

    @Override
    public void abort() {
        aborted = true;
        if (abortDelegate != null)
            abortDelegate.abort();
        super.abort();
    }

    @Override
    protected void execute() throws Exception {
        coreCommand.run();
        processAbortAndPauseEvents();
    }
}

File: CSharp/ExtensibleCommands/ExtensibleCommands/AbortableCommand.cs
using System;

namespace ExtensibleCommands
{
    /// <summary>
    /// Implements command that can be aborted during its execution by invoking Abort() method.
    /// This implies that command execution and its abort should occur in different threads.
    /// </summary>
    public class AbortableCommand : DecoratorCommand
    {
        /// <summary> Type of delegate to abort a command </summary>
        public delegate void AbortDelegate();

        /// <summary> Method to abort this command </summary>
        private readonly AbortDelegate _abortDelegate;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="abortDelegate">Delegate to abort operation of the Core command</param>
        public AbortableCommand(ICommand coreCommand, AbortDelegate abortDelegate) 
            : this(coreCommand, abortDelegate, "Abortable") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="abortDelegate">Delegate to abort the operation of the Core command</param>
        /// <param name="name">Command name</param>
        public AbortableCommand(ICommand coreCommand, AbortDelegate abortDelegate, string name)
            : base(coreCommand, name)
        {
            if (abortDelegate == null)
                throw new Exception(string.Format("Abort Delegate is NULL in AbortableCommand {0}", Name));

            _abortDelegate = abortDelegate;
        }

        public override void Abort()
        {
            _aborted = true;
            _abortDelegate?.Invoke();
            base.Abort();
        }

        protected override void Execute()
        {
            CoreCommand.Run();
            ProcessAbortAndPauseEvents();
        }
    }
}
